### PR TITLE
add BasicOrganizationView, giving membership info to org objects

### PR DESCRIPTION
**Pull request by camhashemi**
@ryanpbrewster @andrewconner we initially talked about keeping the `membership` independent of the `organization` object, but mobile and @cvializ are asking for this now.

### CODE REVIEW

**Comment by cvializ** [comment]
bump @andrewconner @ryanpbrewster

**Comment by andrewconner** [comment]
Good with me.

**Comment by camhashemi** [comment]
@cvializ do we need web-handling for this? will this break things?

**Comment by cvializ** [comment]
I see that you removed orgMembership from the library object, but nothing was using that property anyway. I think it's good, unless I missed another property being removed

File: kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala [hunk 1]
```diff
@@ -88,18 +87,42 @@ object OrganizationInviteInfo {
...
-case class OrganizationView(
+case class BasicOrganizationView(
+  basicOrganization: BasicOrganization,
+  membershipInfo: OrganizationMembershipInfo)
+object BasicOrganizationView {
+  implicit val defaultWrites: Writes[BasicOrganizationView] = new Writes[BasicOrganizationView] {
+    def writes(o: BasicOrganizationView) = Json.toJson(o.basicOrganization).as[JsObject] + ("membership" -> Json.toJson(o.membershipInfo))
+  }
+
+  val testReads = new Reads[BasicOrganizationView] {
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala]: This is just to make tests easier? Do you think it actually makes them easier? I've thought a bit about whether it actually makes sense to test our serialization by making sure that our deserialization works. At this point I actually prefer to directly test the serialized output against what we expect (field by field, not checking the entire JSON blob).

File: kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala [hunk 2]
```diff
@@ -88,18 +87,42 @@ object OrganizationInviteInfo {
...
+      val handle = (json \ "handle").as[OrganizationHandle]
+      val name = (json \ "name").as[String]
+      val description = (json \ "description").as[Option[String]]
+      val avatarPath = (json \ "avatarPath").as[ImagePath]
+      val basicOrg = BasicOrganization(id, ownerId, handle, name, description, avatarPath)
+      val membershipInfo = (json \ "membership").as[OrganizationMembershipInfo](OrganizationMembershipInfo.testReads)
+      JsSuccess(BasicOrganizationView(basicOrg, membershipInfo))
+    }
+  }
+}
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala:102]: I think there may be a clever Play JSON thing that will handle this a bit more cleanly. If I figure it out I'll change this; it's certainly fine for testing (although I still think that deserialization is the wrong way to handle tests).

File: kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala [hunk 3]
```diff
@@ -88,18 +87,42 @@ object OrganizationInviteInfo {
...
+
+case class FullOrganizationView(
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala]: My vote is for `FullOrganizationView` to be named `OrganizationView`

File: kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala (hunk 1 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala]: yeah, I do that too, extracting the fields I expect to be there, but this is more robust than that (you can't add untested fields), without the pain of an entire JSON matching.

File: kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentRequest.scala [hunk 4]
```diff
@@ -17,7 +17,7 @@ object AccountFeatureSettingsRequest {
     def reads(json: JsValue): JsResult[AccountFeatureSettingsRequest] = {
       json.validate[JsObject].map { obj =>
         val featureSettings = obj.fieldSet.map {
-          case (key, value) if Feature.get(key).isDefined && Feature.get(key).get.verify(value.as[String]) => FeatureSetting(key, value.as[String])
+          case (key: String, value: JsValue) if Feature.get(key).isDefined && Feature.get(key).get.verify(value.as[String]) => FeatureSetting(key, value.as[String])
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentRequest.scala:20]: Why are we checking to see if `Feature.get(key).isDefined`? I think we should have `FeatureSettings` being a map from `Feature -> Setting` and then we define a `Feature.apply` method that uses a `match` statement. That is the convention we use elsewhere in the codebase.

Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentRequest.scala:20]: Agreed. I think the original PR has a suggestion on how to do it cleanly.

File: kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala (hunk 2 above)
Comment (maxov) [inline kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala:102]: I think just `(json).as[BasicOrganization]` and `(json \ "membership").as[OrganizationMembershipInfo]`would work fine, I don't think that play json errors with extra fields

Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala:102]: Correct, and that'd definitely work. I think there's an easy way to make Play serialize a single JsObject into a tuple, though (and therefore into a case class that wraps a tuple). Seems slightly cleaner.

File: kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala (hunk 1 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala]: You convinced me. This seems reasonable.

File: kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala (hunk 2 above)
Comment (maxov) [inline kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala:102]: There's no Play tuple format that I know of. We define a special tuple format that serializes and deserializes into json arrays.

Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala:102]: You're right, I guess your way is probably the cleanest.

```
val testReads: Reads[BasicOrganizationView] = (
  __.read[BasicOrganization] and
  (__ \ "membership").read[OrganizationMembershipInfo]
)(BasicOrganizationView.apply _)
```

Comment (maxov) [inline kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala:102]: :+1:

Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/model/OrganizationInfo.scala:102]: nice, that's awesome. thanks!

File: kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentRequest.scala (hunk 4 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentRequest.scala:20]: yup. this is all coming soon.

**Pull request merged**